internal/unionstore/art/art_iterator.go

```diff
@@ -0,0 +1,533 @@
+				return n256.inplaceLeaf
+			}
+		} else if idx == 256 {
+			break
+		}
+		idx = n256.nextPresentIdx(idx)
+		if idx < 256 {
+			it.idxes[depth] = idx
+			child := n256.children[idx]
+			if child.kind == typeLeaf {
+				it.idxes[depth]++
+				return child
+			}
+			it.nodes = append(it.nodes, child)
+			it.idxes = append(it.idxes, -1)
+			return it.next()
+		}
+	}
+	it.nodes = it.nodes[:depth]
+	it.idxes = it.idxes[:depth]
+	if depth == 0 {
+		return nullArtNode
+	}
+	it.idxes[depth-1]++
+	return it.next()
+}
+
+func (it *baseIter) prev() artNode {
+	depth := len(it.nodes) - 1
+	curr := it.nodes[depth]
+	idx := it.idxes[depth]
+	idx--
+	switch curr.kind {
+	case typeNode4:
+		n4 := it.allocator.getNode4(curr.addr)
+		idx = min(idx, int(n4.nodeNum)-1)
+		if idx >= 0 {
+			it.idxes[depth] = idx
+			child := n4.children[idx]
+			if child.kind == typeLeaf {
+				return child
+			}
+			it.nodes = append(it.nodes, child)
+			it.idxes = append(it.idxes, node256cap)
+			return it.prev()
+		} else if idx == -1 {
+			it.idxes[depth] = idx
+			if n4.inplaceLeaf.addr != nullAddr {
+				return n4.inplaceLeaf
+			}
+		} else if idx == -2 {
+			break
+		}
+	case typeNode16:
+		n16 := it.allocator.getNode16(curr.addr)
+		idx = min(idx, int(n16.nodeNum)-1)
+		if idx >= 0 {
+			it.idxes[depth] = idx
+			child := n16.children[idx]
+			if child.kind == typeLeaf {
+				return child
+			}
+			it.nodes = append(it.nodes, child)
+			it.idxes = append(it.idxes, node256cap)
+			return it.prev()
+		} else if idx == -1 {
+			it.idxes[depth] = idx
+			if n16.inplaceLeaf.addr != nullAddr {
+				return n16.inplaceLeaf
+			}
+		} else if idx == -2 {
+			break
+		}
+	case typeNode48:
+		n48 := it.allocator.getNode48(curr.addr)
+
+		if idx >= 0 {
+			idx = n48.prevPresentIdx(idx)
+		}
+
+		if idx >= 0 {
+			it.idxes[depth] = idx
+			child := n48.children[n48.keys[idx]]
+			if child.kind == typeLeaf {
+				return child
+			}
+			it.nodes = append(it.nodes, child)
+			it.idxes = append(it.idxes, node256cap)
+			return it.prev()
+		} else if idx == -1 {
+			it.idxes[depth] = idx
+			if n48.inplaceLeaf.addr != nullAddr {
+				return n48.inplaceLeaf
+			}
+		} else if idx == -2 {
+			break
+		}
+	case typeNode256:
+		n256 := it.allocator.getNode256(curr.addr)
+
+		if idx >= 0 {
+			idx = n256.prevPresentIdx(idx)
+		}
+
+		if idx >= 0 {
+			it.idxes[depth] = idx
+			child := n256.children[idx]
+			if child.kind == typeLeaf {
+				return child
+			}
+			it.nodes = append(it.nodes, child)
+			it.idxes = append(it.idxes, node256cap)
+			return it.prev()
+		} else if idx == -1 {
+			it.idxes[depth] = idx
+			if n256.inplaceLeaf.addr != nullAddr {
+				return n256.inplaceLeaf
+			}
+		} else if idx == -2 {
+			break
+		}
+	}
+	it.nodes = it.nodes[:depth]
+	it.idxes = it.idxes[:depth]
+	if depth == 0 {
+		return nullArtNode
+	}
+	return it.prev()
+}
+
+func (it *baseIter) close() {
+	it.nodes = it.nodes[:0]
+	it.idxes = it.idxes[:0]
+}
+
+type ArtMemKeyHandle struct {
```

Comment:
Use `MemKeyHandle` will lead to cyclic imports now actually.

`ArtMemKeyHandle` is same to `MemKeyHandle` because the mem arena is same, but I think they can be different for different MemBuffer implementation.

The best choice is associated type. But associated type is not supported, maybe try generic type in `memBufferMutations` if we want it to support all the MemBuffer implementation.